### PR TITLE
Add api calls for oauth frontend

### PR DESCRIPTION
**Pull request by Philinphiladelphia**
## Issue Number

#148, #281, #280 

## Purpose/Implementation Notes

This adds the context, api calls, and UI linking for ORCID OAUTH. It adds:
- Changes to the token authentication for the backend
- An updated auth endpoint
- Frontend api calls to authenticate, get user info, and refresh tokens
- Context to store the user info
- Mechanisms to store user info in the local storage
- Showing/hiding the login button based on the presence of a logged-in user

Supporting changes:
- Update the recreate-schemas command to work with networked containers
- Update every single test that has "unauthenticated" in it to use a 401 instead of a 403

NOTE that at this point the UI PR which this depends on is not merged. Once it is the diff will not include certain component files. Nonetheless, this PR got a bit... huge. I apologize for that.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Functional tests

None.

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/resources_portal/views/auth.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                {
+                    "error": f"Origin URL parameter was not found in the URL: {request.build_absolute_uri()}"
+                },
+                status=400,
+            )
+
+        authorization_code = request.GET["code"]
+        origin_url = request.GET["origin_url"]
+
+        print("GET PARAMS: ", request.GET)
```
Comment (kurtwheeler) [inline api/resources_portal/views/auth.py:60]: This looks like leftover debugging code.

File: api/resources_portal/views/auth.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+        # get user orcid info
+        response = requests.post(OAUTH_URL, data=data, headers={"accept": "application/json"})
+        response_json = response.json()
+        if "orcid" not in response_json:
+            return JsonResponse(response_json, status=400)
+
+        user = User.objects.filter(orcid=response_json["orcid"]).first()
+
+        print("USer in database: ", user)
```
Comment (kurtwheeler) [inline api/resources_portal/views/auth.py:78]: There's several print statements in this file.

File: api/resources_portal/views/auth.py (hunk 1 above)
Comment (Philinphiladelphia) [reply api/resources_portal/views/auth.py]: These are all gone now!

File: api/resources_portal/views/auth.py (hunk 2 above)
Comment (Philinphiladelphia) [reply api/resources_portal/views/auth.py]: See above.

File: client/src/hooks/useUser.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      setUser(defaultUser)
+      setToken(defaultToken)
+      if (loginRedirectUrl) {
+        setLoginRedirectUrl()
+        router.replace(loginRedirectUrl)
+      }
+    }
+  })
+  const isLoggedIn = Boolean(user && token)
+  const refreshUserData = () => {
```
Comment (davidsmejia) [inline client/src/hooks/useUser.js:28]: I think this needs to be async and we need to await the api calls

File: client/package-lock.json [hunk 4]
```diff
@@ -1,14229 +0,0 @@
...
-          "requires": {
-            "errno": "^0.1.3",
-            "readable-stream": "^2.0.1"
-          }
-        }
-      }
-    },
-    "entities": {
-      "version": "1.1.2",
-      "resolved"
```
Comment (kurtwheeler) [inline client/package-lock.json:14229]: Was this deletion intentional?

File: api/resources_portal/migrations/0014_auto_20200812_2144.py [hunk 5]
```diff
@@ -12,6 +12,12 @@ class Migration(migrations.Migration):
     ]
 
     operations = [
+        migrations.RenameField(
+            model_name="user", old_name="access_token", new_name="orcid_access_token",
+        ),
+        migrations.RenameField(
+            model_name="user", old_name="refresh_token", new_name="orcid_refresh_token",
+        ),
```
Comment (kurtwheeler) [inline api/resources_portal/migrations/0014_auto_20200812_2144.py:20]: I don't think you can modify old migrations like this. These have already been run on the staging server.

File: api/resources_portal/test/end_to_end/test_organization_with_two_users.py [hunk 6]
```diff
@@ -118,9 +118,10 @@ def test_organization_with_two_users(self, mock_auth_request, mock_record_reques
         # Prof removes all notification settings
         self.client.force_authenticate(user=prof)
 
-        settings_url = reverse(
-            "organization-user-setting-detail", args=[prof.organization_settings.first().id]
-        )
+        # Temporary, I want to submit another PR to make this endpoint accept a user and an org. It makes more sense
```
Comment (kurtwheeler) [inline api/resources_portal/test/end_to_end/test_organization_with_two_users.py:121]: Do you have an issue to track this?

File: api/resources_portal/views/auth.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    """
+    This removes the "code" parameter added by the first ORCID call if it is there,
+     and trims off the trailing '/?' if it is there.
+    """
+    return url.split("code")[0].strip("&").strip("/?")
+
+
+class AuthViewSet(viewsets.ViewSet):
+
+    # Verify that this should be a GET/POST?
```
Comment (kurtwheeler) [inline api/resources_portal/views/auth.py:32]: What's this comment about?

File: client/src/api/index.js [hunk 8]
```diff
@@ -20,12 +20,20 @@ const getAPIURL = (endpoint = '', query = {}) => {
...
+  const { body } = options
   const config = { ...options }
+  config.body = body
   config.headers = {
     'content-type': 'application/json',
     ...headers
   }
+  if (authorization) config.headers.Authorization = `Token ${authorization}`
+
+  console.log('options: ', config)
```
Comment (kurtwheeler) [inline client/src/api/index.js:36]: Did you mean to leave this log here?

File: client/src/api/index.js [hunk 9]
```diff
@@ -44,11 +52,46 @@ const request = async (url, { headers = {}, ...options } = {}) => {
   }
 }
 
+export const loginUser = () => {}
+// return array of api responses
```
Comment (kurtwheeler) [inline client/src/api/index.js:56]: What's this comment about?

**Comment by davidsmejia** [comment]
oh yeah i meant to say that the yarn stuff should be removed from the main folder, it should only be in /client

File: client/src/hooks/useUser.js (hunk 3 above)
Comment (Philinphiladelphia) [reply client/src/hooks/useUser.js:28]: Cool, done.

File: api/resources_portal/test/end_to_end/test_organization_with_two_users.py (hunk 6 above)
Comment (Philinphiladelphia) [reply api/resources_portal/test/end_to_end/test_organization_with_two_users.py:121]: I do now (: https://app.zenhub.com/workspaces/refinebio-5c92300a9d1bd30b8d3101f2/issues/alexslemonade/resources-portal/316

File: api/resources_portal/views/auth.py (hunk 7 above)
Comment (Philinphiladelphia) [reply api/resources_portal/views/auth.py:32]: I was a bit worried about security on this endpoint, but we've now verified that nothing protected is being passed here. Removed.

File: client/src/api/index.js (hunk 8 above)
Comment (Philinphiladelphia) [reply client/src/api/index.js:36]: I did not. Nice catch

File: client/src/api/index.js (hunk 9 above)
Comment (Philinphiladelphia) [reply client/src/api/index.js:56]: A note for myself based on something David was explaining to me.

Comment (Philinphiladelphia) [reply client/src/api/index.js:56]: Removed.

File: api/resources_portal/views/auth.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+                    {
+                        "error": "There is no user associated with the given URL and no 'email' parameter was provided to create one."
+                    },
+                    status=400,
+                )
+
+            email = request.GET["email"]
+
+            # Get first and last name
+            api = orcid.PublicAPI(CLIENT_ID, CLIENT_SECRET, sandbox=True)
```
Comment (davidsmejia) [inline api/resources_portal/views/auth.py:82]: can we make sandbox=True source from a config variable?

**Comment by davidsmejia** [comment]
Looks like you might need to run `yarn upgrade` because some of the packages in the yarn lock were being down versioned. Not sure how that happened though

File: api/resources_portal/config/common.py [hunk 11]
```diff
@@ -195,5 +194,9 @@ class Common(Configuration):
         "guardian.backends.ObjectPermissionBackend",
     )
 
+    # Time for expiration of API tokens
+    # About 6 months
+    EXPIRING_TOKEN_DURATION = timedelta(weeks=26)
```
Comment (kurtwheeler) [inline api/resources_portal/config/common.py:199]: Any reason not to just use `timedelta(months=6)`?

Comment (Philinphiladelphia) [reply api/resources_portal/config/common.py:199]: timedelta's largest available time interval is weeks.

**Pull request merged**